File: sudoku.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <iterator>
#include <time.h>
#include <sys/time.h>

const int BOARD_SIZE = 9;

bool nextEmpty(int sudoku[][BOARD_SIZE], int spot[2]);

bool canPut(int sudoku[][BOARD_SIZE], int spot[2], int val);

void solve(int sudoku [][BOARD_SIZE]);

void fromString(const std::string &data, int board[][BOARD_SIZE]);

std::string toString(int board[][BOARD_SIZE]);


int main(int argc, char const *argv[])
{
	std::ifstream infile(argv[1]);
	if (infile)	 {

		// read entire file into memory
		std::string fileData((std::istreambuf_iterator<char> (infile)) ,
			std::istreambuf_iterator<char>()) ;
		infile.close();
        std::stringstream ssinput (fileData);


		// solve all sudokus
		struct timeval before;
		gettimeofday (&before, NULL);

		std::stringstream ssoutput;
		std::string line;
		while (std::getline (ssinput, line)) {
			int board[BOARD_SIZE][BOARD_SIZE];
			fromString (line, board);
			solve (board);
			ssoutput << toString (board);
		}

		struct timeval after;
		gettimeofday (&after, NULL);
		long duration = ((after.tv_sec * 1000000 + after.tv_usec) -
			(before.tv_sec * 1000000 + before.tv_usec)) / 1000;

		// write solutions to output
		std::ofstream outfile;
		std::string path (argv[1]);
		std::string name;
		if (unsigned found = path.find_last_of ("/\\")) {
			name = "solved_" + path.substr(found+1);
		} else {
			name = std::string ("solved_") + argv[1];
		}
		outfile.open (name.c_str());
		outfile << ssoutput.str();
		outfile.close();

		// time result
		std::cout << "-- Elapsed time: " << duration << " ms." << std::endl;

		return 0;
	}

	return 0;
}


bool nextEmpty(int sudoku[][BOARD_SIZE], int spot[2]) {
	for (int i=0; i<BOARD_SIZE; i++) {
		for (int j=0; j<BOARD_SIZE; j++) {
			if (sudoku[i][j] == 0) {
				spot[0] = i;
				spot[1] = j;
				return true;
			}
		}
	}
	return false;
}


bool canPut(int sudoku[][BOARD_SIZE], int spot[2], int val) {
	int x = spot[0];
	int y = spot[1];
	for (int i=0; i<BOARD_SIZE; i++) {
            // test row
		if (sudoku[i][y] == val) {
			return false;
		}
            // test column
		if (sudoku[x][i] == val) {
			return false;
		}
	}
     // test square
	int a = x-(x%3);
	int b = y-(y%3);
	for (int i=a; i<a+3; i++) {
		for (int j=b; j<b+3; j++) {
			if (sudoku[i][j] == val) {
				return false;
			}
		}
	}

	return true;
}


void solve(int sudoku [][BOARD_SIZE]) {
	int spot[] = {0, 0};
	if (!nextEmpty(sudoku, spot)) {
		return;
	}

	for (int i=1; i<10; i++) {
		if (canPut(sudoku, spot, i)) {
			sudoku[spot[0]][spot[1]] = i;
			solve(sudoku);
			int _[] = {0, 0};
			if (!nextEmpty(sudoku, _)) {
				return;
			}
		}
	}

        sudoku[spot[0]][spot[1]] = 0; // solution not found, backtrack
    }


void fromString(const std::string &data, int board[][BOARD_SIZE]) {
	int i = 0, j = 0;
	for (std::string::const_iterator it=data.begin();
		it!=data.end() && i < BOARD_SIZE && j < BOARD_SIZE;
		++it) {
		if (isdigit(*it)) {
			board[i][j] = *it - 48;
		}
       if (j == BOARD_SIZE - 1) {
            i += 1 % BOARD_SIZE;
            j = 0;
        } else {
            j += 1 % BOARD_SIZE;
        }
	}
}


std::string toString(int board[][BOARD_SIZE]) {
	std::stringstream ss;
	for (int i=0; i<BOARD_SIZE; i++) {
		for (int j=0; j<BOARD_SIZE; j++) {
			ss << board[i][j] << " ";
		}
		ss << "\n";
	}
	ss << "\n";
	return ss.str();
}

File: Sudoku.scala
import java.io.File
import java.util.Scanner
import java.io.BufferedWriter
import java.io.FileWriter

// Build a sudoku from a 81 char line
class Sudoku(text: String) {
	val rowSize = 9
  
	val board = Array.ofDim[Integer](rowSize, rowSize)
	var i = 0
	var j = 0
	for (char <- text) {
		board(i)(j) = char.toInt - 48
		if (j == rowSize - 1) {
			i += 1 % rowSize
			j = 0
		} else {
			j += 1 % rowSize
		}
	}
	
	override def toString:String =
	{
		val sb = new StringBuilder()
		val lineSep = System.getProperty("line.separator")
		for (row:Array[Integer] <- this.board) yield {
		  sb.append((row map (i => i.toString)).mkString(" ")).append(lineSep)
		}
		sb.toString
	}
	
}

object SudokuSolver {
	
	def solve(sudoku:Sudoku):Sudoku =
	{
		val spot = nextEmpty(sudoku)
		spot match {
		  case None => sudoku
		  case Some((i,j)) =>
		    for (value <- 1 to 9) {
		    	if (canPut(sudoku, (i,j), value)) {
		    		sudoku.board(i)(j) = value
		    		val sudoku2 = solve(sudoku)
		    		if (!nextEmpty(sudoku2).isDefined) {
		    			return sudoku2 // solution found
		    		}
		    	}
		    }
		    // backtrack, solution not found
		    sudoku.board(i)(j) = 0
		    return sudoku
		}
	}
	
	def nextEmpty(sudoku:Sudoku):Option[Tuple2[Integer,Integer]] =
	{
		for (i <- 0 to sudoku.rowSize-1) {
			for (j <- 0 to sudoku.rowSize-1) {
				if (sudoku.board(i)(j) == 0) return Some((i,j)) 
			}
		}
		None
	}
	
	def canPut(sudoku:Sudoku, spot:Tuple2[Integer,Integer], value:Integer):Boolean =
	{
		for (i <- 0 to sudoku.rowSize-1) {
			// test row
			if (sudoku.board(i)(spot._2) == value) return false
			
			// test column
			if (sudoku.board(spot._1)(i) == value) return false
		}
		
		// test square
		val a = spot._1 - (spot._1%3)
		val b = spot._2 - (spot._2%3)
		for (i <- a to a+2) {
			for (j <- b to b+2) {
				if (sudoku.board(i)(j) == value) return false
			}
		}
		
		true
	}
}

class ProgressBar(var totalSteps:Integer, var step:Integer, var resolution:Integer, var width:Integer) {
	def advance 
	{
		step += 1
        if (totalSteps/resolution == 0) return
        if (step % (totalSteps/resolution) != 0) return
        var ratio:Float = step/totalSteps.toFloat
        val count:Int = (ratio * width).toInt
        print("%3d%% [".format((ratio * 100).toInt))
		(0 to count-1) foreach {_ => print("=")}
		(count to width-1) foreach {_ => print(" ")}
        print( "]\r ")
        System.out.flush()
	}
}

object Sudoku {

	def main(args: Array[String]) {
	  
		val file = new File(args(0))
		
		// read entire input into memory
		val input = new Scanner(file, "UTF-8").useDelimiter("\\A").next()
		
		val output = new StringBuilder()
		
		val before = System.nanoTime
		
		val lines = input.split("\\r?\\n")
		
		val progressBar = new ProgressBar(0, lines.length, 100, 100)

		val lineSep = System.getProperty("line.separator")
		for (line <- lines) {
			val sudoku = new Sudoku(line)
			output.append(SudokuSolver.solve(sudoku)).append(lineSep)
			progressBar.advance
		}
		
		// elapsed time
		println("\nElapsed time: " + ((System.nanoTime - before)/1e6) + " ms")
		
		// write solutions to file
		val writer = new BufferedWriter(new FileWriter("solved_" + file.getName()))
		writer.write(output.toString)
		writer.close
	}
}